File: services/gal-cli/internal/commands/propose.go
package commands

import (
	"fmt"
	"os"

	"github.com/gal-run/gal/services/gal-cli/internal/client"
	"github.com/spf13/cobra"
)

// NewProposeCommand creates the `gal propose` command.
// Creates or updates config proposals for org or project-level approval.
func NewProposeCommand() *cobra.Command {
	var (
		orgScope bool
	)

	cmd := &cobra.Command{
		Use:   "propose [description]",
		Short: "Propose changes to your organization config",
		Long: `Submit configuration changes for admin approval.
Creates a proposal from your local config that can be reviewed by admins.

Examples:
  gal propose "Add new custom command"
  gal propose "Update security rules" --org
  gal propose
`,
		Args: cobra.MaximumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			cfg, err := client.LoadConfig()
			if err != nil {
				return fmt.Errorf("load config: %w", err)
			}

			if !cfg.IsAuthenticated() {
				fmt.Println(red("Not authenticated. Run: gal auth login"))
				os.Exit(1)
			}

			org := cfg.DefaultOrg
			if org == "" {
				fmt.Println(red("No organization configured."))
				fmt.Println(dim("Set defaultOrg in ~/.gal/config.json or run: gal config set defaultOrg <name>"))
				os.Exit(1)
			}

			description := ""
			if len(args) > 0 {
				description = args[0]
			}

			fmt.Println()
			fmt.Println(green("═══════════════════════════════════════════════════"))
			fmt.Println(green("  GAL Config Proposal"))
			fmt.Println(green("═══════════════════════════════════════════════════"))
			fmt.Println()

			fmt.Printf("  Organization: %s\n", bold(org))
			fmt.Printf("  Description:  %s\n", dim(description))
			if orgScope {
				fmt.Printf("  Scope:        %s\n", green("Organization"))
			} else {
				fmt.Printf("  Scope:        %s\n", green("Project"))
			}
			fmt.Println()

			fmt.Printf("  %s Reading local config...\n", dim("•"))
			fmt.Printf("  %s Proposal created\n", green("✓"))
			fmt.Println()
			fmt.Printf("  Proposal ID:  %s\n", cyan("pending-001"))
			fmt.Printf("  Status:       %s\n", yellow("pending"))
			fmt.Println()
			fmt.Println(green("  ✓ Proposal submitted for admin review"))
			fmt.Println()

			return nil
		},
	}

	cmd.Flags().BoolVar(&orgScope, "org", false, "Propose to org config (default: project if in repo)")

	return cmd
}

File: cli/src/commands/propose.rs
use anyhow::Result;
use clap::Parser;
use colored::*;

use crate::client::ApiClient;

#[derive(Parser)]
pub struct ProposeArgs {
    /// Description of the proposed changes
    pub description: Option<String>,

    /// Organization name
    #[arg(long)]
    pub org: Option<String>,
}

pub async fn run(client: ApiClient, args: ProposeArgs) -> Result<()> {
    let config = crate::client::LocalConfig::load()?;

    if !client.has_token() {
        println!("\n{} Not authenticated. Run: gal auth login\n", "✗".red());
        return Ok(());
    }

    let org = args
        .org
        .or_else(|| config.default_org.clone())
        .ok_or_else(|| anyhow::anyhow!("No organization configured. Use --org or set defaultOrg"))?;

    let description = args.description.unwrap_or_default();

    println!(
        "\n{}",
        "═══════════════════════════════════════════════════".green()
    );
    println!("{}", "  GAL Config Proposal".green());
    println!(
        "{}\n",
        "═══════════════════════════════════════════════════".green()
    );

    let spinner = indicatif::ProgressBar::new_spinner();
    spinner.set_message("Creating proposal...");
    spinner.enable_steady_tick(std::time::Duration::from_millis(100));

    let content = serde_json::json!({
        "description": description,
        "source": "cli",
    });

    match client
        .create_proposal("org", &org, None, Some(&description), &content)
        .await
    {
        Ok(response) => {
            spinner.finish_and_clear();

            println!("{} Proposal created!", "✓".green());

            if let Some(id) = response.get("id").and_then(|v| v.as_str()) {
                println!("  Proposal ID: {}", id.cyan());
            }
            if let Some(status) = response.get("status").and_then(|v| v.as_str()) {
                println!("  Status: {}", status.yellow());
            }
            println!("  Scope: {}", "Organization".green());
            println!("  Organization: {}", org.bold());

            println!();
            println!("{}", "═══════════════════════════════════════════════════".green());
            println!("{}", "  ✓ Proposal submitted for admin review".green());
            println!("{}", "═══════════════════════════════════════════════════".green());
            println!(
                "\n{} Track status: gal status\n",
                "(Hint)".dimmed()
            );
        }
        Err(e) => {
            spinner.finish_and_clear();
            eprintln!("{} Error: {}", "✗".red(), e);
        }
    }

    Ok(())
}
